### ISSUE TITLE
Frequency of Location Events Too Slow: 1 location recorded for a 5 minute / 1 mile drive

### ISSUE THREAD

**Issue by josiahbryan** [mauron85/cordova-plugin-background-geolocation#626]
## Your Environment

* Plugin version: 3.0.6
* Platform: Android
* OS version: 10
* Device manufacturer and model: Google Pixel 3

* Running in Simulator: No
* Cordova version (`cordova -v`): PhoneGap Build (Cordova cli 9.0)
* Cordova platform version (`cordova platform ls`): PhoneGap build...?
* Plugin configuration options: 
        <variable name="GOOGLE_PLAY_SERVICES_VERSION" value="11.0.1" />
        <variable name="ANDROID_SUPPORT_LIBRARY_VERSION" value="23+" />
        <variable name="ALWAYS_USAGE_DESCRIPTION" value="App requires background tracking enabled" />
        <variable name="MOTION_USAGE_DESCRIPTION" value="App requires motion detection" />
* Link to your project: https://walksafe.app (app store publish pending), you can find the apk on PhoneGap build at https://build.phonegap.com/apps/3486871/install/yAxPGgfqaUW6sxfVXW7b

## Context
I have the plugin configured for 1000ms intervals, with ACTIVITY provider (activity interval is 2500). However, on a 1-mile drive (walk to car, drive, walk to my desk), I only received a single background location event. One event. I would think over a one mile drive (roughly 5 minutes), there should have been at least 2 or 3, ideally 10 or 20 events...yes? Am I wrong, my config wrong, or the code wrong? LoL!

Here's my config:
```
		BackgroundGeolocation.configure({
			locationProvider:   BackgroundGeolocation.ACTIVITY_PROVIDER,
			desiredAccuracy:    BackgroundGeolocation.HIGH_ACCURACY,
			stationaryRadius:   10,
			distanceFilter:     15,
			notificationTitle:  'WalkSafe is keeping you safe',
			notificationText:   'WalkSafe tracks your location and speed to see if you might be in danger and to log your location for your emergency contacts when sending an SOS message',
			debug:              false,
			interval:           1000,
			fastestInterval:    500,
			activitiesInterval: 2500,
			startOnBoot:        true,
			stopOnTerminate:    false,
			url:                server.urlRoot + '/walksafe/store_gps_location',
			httpHeaders: {
				// We should have token by now because
				// connectNativeLogger() only called by LOGIN_EVENT
				'Authorization': server.token,
			},
			// customize post properties
			postTemplate: {
				lat: '@latitude',
				lng: '@longitude',
				accuracy: '@accuracy',
				// Added to attempt server-side walk/run/drive detection
				speed:    '@speed',
				altitude: '@altitude',
				bearing:  '@bearing',
				// built above from deviceInfo()
				sensor
			}
		});
```  

## Expected Behavior
I should have frequent background location events (more than 1 a minute)

## Actual Behavior
Only a single background location posted over 5 minutes on a 1 mile drive

## Possible Fix
Option to remove any time limitations or artificial restrictions on post frequency?

## Steps to Reproduce
1. Use config above
2. Load into an app
3. Go for a 1 mile drive
4. Check server
5. Look sadly at the screen and see only one or two event 

## Context
Trying to record location to keep users safe

## Debug logs
No logs available. I can claw them out of the device if requested.

**Comment by mauron85** [comment]
>Option to remove any time limitations or artificial restrictions on post frequency?

there are no post limitation. Try to enable `debug: true` and listen to location sounds (how frequently they occur) or [check log entries](https://github.com/mauron85/cordova-plugin-background-geolocation#getlogentrieslimit-fromid-minlevel-success-fail).

**Comment by josiahbryan** [comment]
TL;DR: Guessing a crash, here's logs, my app is `com.josiahbryan.walksafe`: [logcat.txt](https://github.com/mauron85/cordova-plugin-background-geolocation/files/3621827/logcat.txt)

Okay, crashing, I think ... I tried on my drive to work to test with debug: true - I didn't see your edits till just now, sorry - will totally try those too.

But, back to crashing. I had the app in the background with `debug: true` running for my hour-long drive to the office. (Yuck. But that's another story.)

So, what I experienced was indeed a bunch of toasts and beeps - then they would stop after roughly 20 minutes. So I would go to the app - which was still active in the background (e.g. showing recent apps and touching the app brought it instantly to the foreground, there was no "reloading" state change where the webview had to boot). 

In order to get the toasts and beeps to start again, I had to "swipe up" to close the app (Pixel 3), then open the app again. As soon as I opened the app again from the launcher, beeps would start, toasts would show....then stop after roughly 20 minutes. I swiped to close and opened once more when I was walking to my office after parking. And it happened again - lots of beeps, then stopped about 20 minutes later.

So, once I got to my desk, I plugged in and `adb logcat` dumped to a file. That file is attached. My app's package name is `com.josiahbryan.walksafe`, and a search thru the logs revealed mostly useless stuff (for my eyes anyway), but this line stood out to me:

`09-17 09:51:32.409  1345  1384 W ActivityManager: Stopping service due to app idle: u0a378 -7m42s239ms com.josiahbryan.walksafe/com.marianhello.bgloc.service.LocationServiceImpl`

And a little while later, these lines, but to be fair, this is about when I might have killed it myself, not sure:
```09-17 09:53:22.278  7032  7032 W PrimesInit: Primes instant initialization
09-17 09:53:22.279  7032  7069 W Primes  : background initialization
09-17 09:53:22.319  7032  7069 I PrimesTesting: GserviceFlagsSupplier.get()
09-17 09:53:22.389  1345  3081 I ActivityManager: Killing 2601:com.josiahbryan.walksafe:sync/u0a378 (adj 985): empty #17
09-17 09:53:22.426   846   846 I Zygote  : Process 2601 exited due to signal 9 (Killed)
09-17 09:53:22.435  2496  2496 W ChimeraUtils: Non Chimera context
09-17 09:53:22.435  2496  2496 W ChimeraUtils: Non Chimera context
09-17 09:53:22.436  1345  1386 I libprocessgroup: Successfully killed process cgroup uid 10378 pid 2601 in 45ms
```

Any thoughts on any of that or my logs? More than happy to change config and re-test, just need advice as to what to try. Thanks!

**Comment by josiahbryan** [comment]
Tried adding your edits in from above, and reloaded my app. Nothing really changed, if anything, less debugging alerts. To be fair, I also changed some timeouts. However, what stood out to me in the new logs was this:

```09-17 11:08:50.248  1345  1384 W ActivityManager: Stopping service due to app idle: u0a378 -22m36s344ms com.josiahbryan.walksafe/com.marianhello.bgloc.service.LocationServiceImpl
09-17 11:08:50.256  2083  2083 I com.marianhello.bgloc.service.LocationServiceImpl: Destroying LocationServiceImpl
09-17 11:08:50.265  2083  2083 I com.marianhello.bgloc.provider.ActivityRecognitionLocationProvider: Destroying ActivityRecognitionLocationProvider
09-17 11:08:50.279  2083  2083 I com.marianhello.bgloc.provider.ActivityRecognitionLocationProvider: Stop recording
09-17 11:08:50.279  2496 20998 I LockboxService: Completed task: true
09-17 11:08:50.282  1345  3081 I BatteryStatsService: In wakeup_callback: suspend aborted
09-17 11:08:50.287  2083  2083 D com.marianhello.bgloc.provider.ActivityRecognitionLocationProvider: Detaching recorder```

Logs from this test attached:
[logcat2.txt](https://github.com/mauron85/cordova-plugin-background-geolocation/files/3622119/logcat2.txt)

And FWIW, new config:
```BackgroundGeolocation.configure({
	locationProvider:   BackgroundGeolocation.RAW_PROVIDER, //ACTIVITY_PROVIDER,
	desiredAccuracy:    BackgroundGeolocation.HIGH_ACCURACY,
	stationaryRadius:   1,
	distanceFilter:     2,
	notificationTitle:  'WalkSafe is keeping you safe',
	notificationText:   'WalkSafe tracks your location and speed to see if you might be in danger and to log your location for your emergency contacts when sending an SOS message',
	debug:              true,
	interval:           1000,
	fastestInterval:    1000,
	activitiesInterval: 1000,
	startOnBoot:        true,
	stopOnTerminate:    false,
	url:                server.urlRoot + '/walksafe/store_gps_location',
	syncUrl:            server.urlRoot + '/walksafe/store_gps_location',
	syncThreshold:      1,
	httpHeaders: {
		// We should have token by now because
		// connectNativeLogger() only called by LOGIN_EVENT
		'Authorization': server.token,
	},
	// customize post properties
	postTemplate: {
		lat:      '@latitude',
		lng:      '@longitude',
		accuracy: '@accuracy',
		// Added to attempt server-side walk/run/drive detection
		speed:    '@speed',
		altitude: '@altitude',
		bearing:  '@bearing',
		// built above from deviceInfo()
		sensor
	}
});```

**Comment by josiahbryan** [comment]
Another update - now it's crashing every time I reinstall the app the first time with this in the logs:

```09-17 12:01:04.990  3852  3852 E com.marianhello.bgloc.service.LocationServiceImpl: processCommand: exceptionjava.lang.SecurityException: Permission Denial: startForeground from pid=3852, uid=10382 requires android.permission.FOREGROUND_SERVICE
```

Which is odd, because, my PhoneGap Build `config.xml` has in it (for cli-9.0.0):

```xml
 <edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest">
        <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
 </edit-config>
```

Full log of crash:
[crash.txt](https://github.com/mauron85/cordova-plugin-background-geolocation/files/3622395/crash.txt)

My config.xml for building: 
[config.xml.txt](https://github.com/mauron85/cordova-plugin-background-geolocation/files/3622399/config.xml.txt)

My log from this latest PhoneGap build:
[build.phonegap.txt](https://github.com/mauron85/cordova-plugin-background-geolocation/files/3622400/build.phonegap.txt)

**Comment by mauron85** [comment]
I will take me some time to "bite" through all your logs, but regarding the last one `requires android.permission.FOREGROUND_SERVICE`, check https://github.com/mauron85/cordova-plugin-background-geolocation/issues/623

Fix is already in git, waiting to published to npm. In meantime just add permission as described in the issue.

**Comment by mauron85** [comment]
Ok, reading your logs, that adding that permission should resolve all your problems.

**Comment by mauron85** [comment]
BTW, just publish fixed version 3.0.7

**Comment by josiahbryan** [comment]
@mauron85 does 3.0.7 include the permission? 👍

**Comment by mauron85** [comment]
yes

**Comment by josiahbryan** [comment]
Dope. And you were right, adding
```xml
 <edit-config file="AndroidManifest.xml" mode="merge" target="/manifest/uses-permission" xmlns:android="http://schemas.android.com/apk/res/android">
          <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
        </edit-config>
```
...did indeed fix it. 

After a short time, the "long running" notification appeared and it kept running.

So, a big huge THANK YOU to your AWESOME support here haha!

Two followups:

- I rebooted my phone, and updates to the server stopped until I restarted the app. Note `startOnBoot` IS `true` in my config ... is that a bug or did I do something wrong in my config?

- Is there any way to get long running background updates without the long running notification? I seem to recall a "for-pay" plugin that seems to do that sans notification, but not sure if that's still feasible or if the Android API requirements have changed...

**Comment by mauron85** [comment]
>I rebooted my phone, and updates to the server stopped until I restarted the app. Note startOnBoot IS true in my config ... is that a bug or did I do something wrong in my config?

well, it should work, try to post adb logs after reboot.

>Is there any way to get long running background updates without the long running notification? I 

this is many times requested feature, but short answer is no. Read more about it in README.md or also https://github.com/mauron85/react-native-background-geolocation/issues/400
